apps/actions-dependencies-updater/src/updater.mts

```diff
@@ -0,0 +1,192 @@
+  [key: string]: {
+    references: {
+      file: string;
+      ref: string;
+    }[];
+    newVersion: {
+      sha: string;
+      version: string;
+    };
+  };
+}
+
+export async function compileUpdates(
+  ctx: RunContext,
+  workflowsByName: WorkflowByName,
+) {
+  // iterate over workflows
+  for (const workflow of Object.values(workflowsByName)) {
+    log.debug(`Workflow: ${workflow.name} - path: ${workflow.path}`);
+
+    for (const job of workflow.jobs) {
+      log.debug(`Job: ${job.name}`);
+
+      for (const dependency of job.directDependencies) {
+        await processActionDependency(ctx, dependency, workflow.path);
+      }
+    }
+  }
+}
+
+async function processActionDependency(
+  ctx: RunContext,
+  action: Action,
+  filePath: string,
+) {
+  if (action.identifier.startsWith("./")) {
+    log.debug(`Local Action: ${action.identifier}`);
+    const actionPath = await getActionYamlPath(
+      join(ctx.repoDir, action.identifier),
+    );
+
+    if (!actionPath) {
+      return;
+    }
+
+    for (const identifier of action.dependencies) {
+      const actionDependency = ctx.actionsByIdentifier[identifier];
+      if (!actionDependency) {
+        log.warn(`Action dependency not found: ${identifier} - skipping.`);
+        continue;
+      }
+      processActionDependency(ctx, actionDependency, actionPath);
+    }
+
+    return;
+  }
+
+  const details = extractDetailsFromActionIdentifier(action.identifier);
+  if (!details) {
+    log.warn(`Unexpected action identifier: ${action.identifier} - skipping.`);
+    return;
+  }
+
+  const { owner, repo, repoPath, ref } = details;
+
+  log.debug(`Checking ${owner}/${repo}${repoPath}@${ref}`);
+
+  const currentVersion = await github.getVersionFromSHA(
+    ctx,
+    owner,
+    repo,
+    repoPath,
+    ref,
+  );
+  const latestVersion = github.getLatestVersion(ctx, owner, repo, repoPath);
+
+  if (!latestVersion) {
+    log.debug(`No newest version found for ${owner}/${repo}${repoPath}@${ref}`);
+    return;
+  }
+
+  if (ref !== latestVersion.sha) {
+    log.debug(
+      `Outdated dependency: ${owner}/${repo}@${ref} => ${currentVersion} < ${latestVersion.version}`,
+    );
+
+    if (currentVersion !== latestVersion.version) {
+      log.debug(
+        `Saving update ${owner}/${repo}${repoPath} from ${currentVersion} to ${latestVersion.version}`,
+      );
+      saveUpdateTransaction(
+        ctx,
+        owner,
+        repo,
+        filePath,
+        ref,
+        latestVersion.sha,
+        latestVersion.version,
+        repoPath,
+      );
+    }
+  }
+}
+
+function saveUpdateTransaction(
+  ctx: RunContext,
+  owner: string,
+  repo: string,
+  filePath: string,
+  existingRef: string,
+  newRef: string,
+  newVersion: string,
+  innerRepoPath?: string,
+) {
+  const cacheKey = `${owner}/${repo}${innerRepoPath || ""}`;
+
+  if (!ctx.caches.updateTransactions.exists(cacheKey)) {
+    ctx.caches.updateTransactions.set(cacheKey, {
```

Comment:
Should we consider a `setIfExists` cache method?

Reply:
Added `getValueOrDefault` - ac12d5a30f1e83fcaa0172b19487b7e25316d4b2